cmd/api/src/api/v2/saved_queries.go

```diff
@@ -242,3 +244,52 @@ func (s Resources) DeleteSavedQuery(response http.ResponseWriter, request *http.
 		response.WriteHeader(http.StatusNoContent)
 	}
 }
+
+// UnshareSavedQueryRequest represents the payload sent to the unshare endpoint
+type UnshareSavedQueryRequest struct {
+	UserIds []uuid.UUID `json:"user_ids"`
+	Self    bool        `json:"self"`
+}
+
+// UnshareSavedQuery allows an owner of a shared query, a user that has a saved query shared to them, or an admin, to remove sharing privileges.
+// A user who owns a query may unshare a query from anyone they have shared to
+// A user who had a query shared to them may unshare that query from themselves
+// And admins may unshare queries that have been shared to other users
+func (s Resources) UnshareSavedQuery(response http.ResponseWriter, request *http.Request) {
+	var (
+		rawSavedQueryID = mux.Vars(request)[api.URIPathVariableSavedQueryID]
+		unshareRequest  UnshareSavedQueryRequest
+	)
+
+	if user, isUser := auth.GetUserFromAuthCtx(ctx2.FromRequest(request).AuthCtx); !isUser {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, "No associated user found", request), response)
+	} else if savedQueryID, err := strconv.ParseInt(rawSavedQueryID, 10, 64); err != nil {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, api.ErrorResponseDetailsFromMalformed, request), response)
+	} else if err := json.NewDecoder(request.Body).Decode(&unshareRequest); err != nil {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, err.Error(), request), response)
+	} else {
+		if unshareRequest.Self {
+			// User is trying to unshare from themselves
+			if err := s.DB.DeleteSavedQueryPermissionsForUser(request.Context(), savedQueryID, user.ID); err != nil {
+				api.HandleDatabaseError(request, response, err)
+			}
+		} else {
+			// Unshare from users in user_ids
+			isAdmin := user.Roles.Has(model.Role{Name: auth.RoleAdministrator})
+			if savedQueryBelongsToUser, err := s.DB.SavedQueryBelongsToUser(request.Context(), user.ID, savedQueryID); err != nil {
+				api.HandleDatabaseError(request, response, err)
```

Comment:
If they are admin we can skip this db hit right?

Reply:
Yeah, I can re-arrange that `if` block